### ISSUE TITLE
BeforeAll/AfterAll-Callbacks not invoked for @ExtendWith annotation on test method

### ISSUE THREAD

**Issue by flok32** [junit-team/junit5#1176]
## Overview

Callbacks for `BeforeAllCallback` and `AfterAllCallback` not running when `@ExtendWith` is not at class level

- **Bug report.**

For the test:
```java
public class MyTest {
    @ExtendWith(MyLifecycle.class)
    @Test
    void test01() {}
}
```
and `MyLifecycle` implementing `BeforeAllCallback` and `AfterAllCallback` the `beforeAll` and `afterAll` callbacks are not invoked.

As the developer of the extension I would expect those callbacks to run even if the `@ExtendWith` annotation is not at class level. Especially if I want to create my own test annotation that uses `@ExtendWith` as meta annotation. Also, the javadoc for the callback methods states that they will be invoked before/after all tests in the current container.

junit version: 5.0.2
junit platform version: 1.0.2
execution environmen: eclipse oxygen 4.7.1a

**Comment by rumpfc** [comment]
I think this is intended. Before-/AfterAll callbacks should be used to setup your test environment which should be used by all of your test methods (not just by a single test method) in your class.

Take a look at the user guide at section [Relative Execution Order of User Code and Extensions](http://junit.org/junit5/docs/current/user-guide/#extensions-execution-order). In the diagram you see a grey box. These methods/callbacks are used before/after each test method when used to extend either a test class or test method.

Outside of this box you see all methods/callbacks which are used before/after **all** test methods when used to extend a test class.

My suggestion: use Before-/AfterAll callbacks as test class extension and not as test method extension

**Comment by sbrannen** [comment]
As stated in the [User Guide](http://junit.org/junit5/docs/current/user-guide/#extensions-registration-inheritance):

> Registered extensions are inherited within test class hierarchies with top-down semantics. Similarly, extensions registered at the class-level are inherited at the method-level.

Thus, "inheritance" does not go in the other direction: a parent (i.e., a _container_) cannot inherit from a child (i.e., a _test_).

**Comment by sbrannen** [comment]
> As the developer of the extension I would expect those callbacks to run even if the `@ExtendWith` annotation is not at class level. 

That would only make sense if the test class only contained one single test method.... or if every single test method registered the same container-level extensions. But in the case of the latter, one would simply register all necessary class-level extensions once at the class level, thereby adhering to the DRY principle.

**Comment by sbrannen** [comment]
> Also, the javadoc for the callback methods states that they will be invoked before/after all tests in the current container.

That's true; however, the implication is that such an extension is actually registered for the container.

That's perhaps a bit vague in the JavaDoc, so we'll improve the wording there.

**Comment by sbrannen** [comment]
FYI: The JavaDoc has been improved in commit 5a3d4fe5916ec07686795019c7368ff7ed29c5e1.
